File: 2021/23.py
cups = [7, 8, 4, 2, 3, 5, 9, 1, 6]
#cups = [3, 8, 9, 1, 2, 5, 4, 6, 7]

cups = cups + [x for x in range(10, 1000001)]

cup_next = dict()
for i in range(len(cups)):
    cup_next[cups[i]] = cups[(i+1) % len(cups)]

cup_prev = dict()
for i in range(len(cups)):
    cup_prev[cups[i]] = cups[(i-1) % len(cups)]


import time
start_time = time.time()

current = cups[0]
for i in range(10000000):
    pickup_1 = cup_next[current]
    pickup_2 = cup_next[pickup_1]
    pickup_3 = cup_next[pickup_2]

    destination_order = [(x - 1) % len(cups) + 1 for x in range(current - 1, current - 5, -1)]

    dest = 0
    for d in destination_order:
        if d == pickup_1 or d == pickup_2 or d == pickup_3:
            next
        else:
            dest = d
            break

    old_pickup_1_prev = cup_prev[pickup_1]
    
    cup_prev[cup_next[dest]] = pickup_3
    cup_prev[cup_next[pickup_3]] = cup_prev[pickup_1]
    cup_prev[pickup_1] = dest

    cup_next[old_pickup_1_prev] = cup_next[pickup_3]
    cup_next[pickup_3] = cup_next[dest]
    cup_next[dest] = pickup_1

    current = cup_next[current]


#print(cup_next)
print(cup_next[1] * cup_next[cup_next[1]])
print("--- %s seconds ---" % (time.time() - start_time))

    

File: 2021/23.R
label <- "784235916"

cups <- as.character(strsplit(label, "")[[1]])
cups <- c(cups, 10:1e6)

cup_list <- as.list(c(cups[2:1e6], cups[1]))
names(cup_list) <- cups

cup_prev <- as.list(c(cups[1e6], cups[1:(1e6-1)]))
names(cup_prev) <- cups

# Dummy -------------------------------------------------------------------


# label <- "389125467"
# 
# cups <- as.character(strsplit(label, "")[[1]])
# 
# cup_list <- as.list(c(cups[2:9], cups[1]))
# names(cup_list) <- cups
# 
# cup_prev <- as.list(c(cups[9], cups[1:8]))
# names(cup_prev) <- cups

# ... ---------------------------------------------------------------------

current <- names(cup_list)[1]

start <- Sys.time()

for(i in 1:1e7) {
    pickup_1 <- cup_list[[current]]
    pickup_2 <- cup_list[[pickup_1]]
    pickup_3 <- cup_list[[pickup_2]]
    
    current_cup <- as.numeric(current)
    destination_order <- ((current_cup - 1):(current_cup - 4) - 1) %% 9 + 1
    destination_cup <- as.character(destination_order[which(!(as.character(destination_order) %in% c(pickup_1, pickup_2, pickup_3)))[1]])

    z <- cup_prev[[pickup_1]]
    
    cup_prev[[cup_list[[destination_cup]]]] <- pickup_3
    cup_prev[[cup_list[[pickup_3]]]] <- cup_prev[[pickup_1]]
    cup_prev[[pickup_1]] <- destination_cup
    
    cup_list[[z]] <- cup_list[[pickup_3]]
    cup_list[[pickup_3]] <- cup_list[[destination_cup]]
    cup_list[[destination_cup]] <- pickup_1
    
    current <- cup_list[[current]]
} 

print(Sys.time() - start)



# print_list(cup_list)

print_list <- function(cup_list) {
    z <- names(cup_list)[1]
    s <- z
    for(i in 2:length(cup_list)) {
        z <- cup_list[[z]]
        s <- sprintf("%s%s", s, z)
    }
    s
}